torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -2738,94 +2738,109 @@ void ProcessGroupGloo::monitoredBarrier(
   if (rank != 0) {
     auto sendWork = send(commTensor, 0, t1);
     auto recvWork = recv(commTensor, 0, t2);
-    sendWork->wait();
-    recvWork->wait();
+    try {
```

Comment:
How about for the following case?

rank 0 received msgs from all ranks first, then send msgs to rank 1 and rank 2, rank1 recvs the msg and will exit monitoredTimeout successfully first, later on rank2 failed to recv the msg, so rank2 will throw errors.  

in this case,  not all ranks failed in monitored barrier, right?

Reply:
In general, if rank != 0, and there is a failure, it will fail before monitored barrier which is the main use case we aim to support, i.e. it will fail before `send()`. In this case, the PR makes it so that all ranks pass or none of them do. 

If all ranks are able to process `send()` but then one fails in `recv()`, it implies that the rank did indeed reach `monitored_barrier` but then somehow errored during it (closed connection, etc). In this case I think it's fine for rank 2 to throw errors (which will be reported by 0) and rank 1 to continue. 

There also doesn't seem to be a great way to solve this case in general, since any subsequent communication could also fail in the same way.

Reply:
that makes sense for major use cases of the monitored barrier, but the doc said 'Note that this collective is all or nothing, in that all ranks will pass the monitored barrier or none of them will.'==> do you want to refactor the doc a little bit? otherwise, the codes look good to me. thanks for the change!

Reply:
That's a good point, will update the doc appropriately as it could be misleading in the case you pointed out.